### ISSUE TITLE
Memory Leak in executemany

### ISSUE THREAD

**Issue by bcmyguest** [psycopg/psycopg#875]
Version: 3.2.1 (binary/c)
Summary: Running `executemany` with `returning=True` seems to have issues where the returned values are not de-allocatedafter the connection and cursor are closed. 

I'm doing a pretty simple write using execute many that needs to return the count of rows that were updated. As far as I can tell, there doesn't seem to be a way for me to get back the free memory that has been allocated to store `_results` in the `BaseCursor`. In my process this seems to consume quite of a bit of memory that doesn't seem to be easily free-able.

Sample code:

```python
statement = sql.SQL(
    """
    INSERT INTO {table} ({columns})
        VALUES ({values})
    ON CONFLICT ({unique})
    DO UPDATE SET {set}
    RETURNING (xmax = 0)
    """
).format(
    table=sql.Identifier("test_table"),
    columns=sql.SQL(",").join([sql.Identifier(c) for c in ["id"]]),
    unique=sql.SQL(",").join([sql.Identifier(c) for c in ["id"]]),
    values=sql.SQL(",").join(
        [sql.Placeholder(c)for c in ["id"]]
    ),
    set=sql.SQL(",").join([sql.Identifier(k) + sql.SQL("=EXCLUDED.") + sql.Identifier(k) for k in ["id"]]),
)

items = [{"id": x} for x in range(1_000_00)]  # large enough to be obvious

conn = psycopg.Connection.connect("url")

try:
    with conn.cursor() as cur:
        cur.executemany(statement, items, returning=True)  # increases memory usage by ~300MB to store results
        conn.commit()
except Exception as e:
    conn.rollback()
    raise e
finally:
    conn.close()
# memory not freed here (even if conn/cur goes out of scope)
``` 

Corrected

**Comment by dvarrazzo** [comment]
I don't see you closing the cursor. If you do (by calling `cur.close()` explicitly or using a `with conn.cursor() as cur` block) memory will be deallocated via a `BaseCursor._reset()` call.

https://github.com/psycopg/psycopg/blob/be4f63915188888d3748fee63c46cbc0e838acff/psycopg/psycopg/_cursor_base.py#L69-L71

**Comment by dvarrazzo** [comment]
Actually I missed it, you do use `with conn.cursor() as cur:`. However this, on exit, calls `BaseCursor._close()`

https://github.com/psycopg/psycopg/blob/be4f63915188888d3748fee63c46cbc0e838acff/psycopg/psycopg/cursor.py#L52-L64

which calls `_reset()`

https://github.com/psycopg/psycopg/blob/be4f63915188888d3748fee63c46cbc0e838acff/psycopg/psycopg/_cursor_base.py#L618-L623

The following script prints `10` and `0`:

```python
import psycopg

with psycopg.connect() as conn:
    with conn.cursor() as cur:
        cur.executemany("select %s", [(i,) for i in range(10)], returning=True)
        print(len(cur._results))
    print(len(cur._results))
```

so maybe your leak is somewhere else?

**Comment by bcmyguest** [comment]
That script is the only script that's needed. The memory for the data in results never gets de-allocated, the script showing the print results doesn't prove much because it's likely there's another pointer that's still referencing the results and causing the issue.

**Comment by dvarrazzo** [comment]
With your error, adding prints at the end of the script:
```
(.venv) piro@sherekhan:~/dev/psycopg3$ python test_leak.py 
cur.closed=False
len(cur._results)=100000
```

without the extra `cur = conn.cursor()`:

```
(.venv) piro@sherekhan:~/dev/psycopg3$ python test_leak.py 
cur.closed=True
len(cur._results)=0
```

**Comment by bcmyguest** [comment]
Yes, you're right I am clobbering the cursor - however removing that does not fix the memory problem. Look at memory usage once the cursor is closed, you will see data not being deallocated correctly.

**Comment by dvarrazzo** [comment]
As far as I could see, adding a print:

```diff
diff --git a/psycopg/psycopg/pq/pq_ctypes.py b/psycopg/psycopg/pq/pq_ctypes.py
index 8335f87f..4450bed0 100644
--- a/psycopg/psycopg/pq/pq_ctypes.py
+++ b/psycopg/psycopg/pq/pq_ctypes.py
@@ -858,6 +858,7 @@ class PGresult:
     def clear(self) -> None:
         self._pgresult_ptr, p = None, self._pgresult_ptr
         if p:
+            print("PQclear")
             PQclear(p)
 
     @property
```

and forcing `PSYCOPG_IMPL=python`, the list is correctly deallocated and the destructor called on every result.

However looking at the allocation from the outside, using ps, I don't see the memory decreasing. For instance:

```
$ ps -o pid,vsz,rss,args -p 2690315
    PID    VSZ   RSS COMMAND
2690315  76884 58548 python test_leak.py    # before executemany

$ ps -o pid,vsz,rss,args -p 2690315
    PID    VSZ   RSS COMMAND
2690315 420140 401460 python test_leak.py    # after executemany

$ ps -o pid,vsz,rss,args -p 2690315
    PID    VSZ   RSS COMMAND
2690315 410140 391852 python test_leak.py    # before close
```

so maybe there's something there.

**Comment by bcmyguest** [comment]
Pardon me for not mentioning it. Yes the cursor clears the results but the impact here can indeed be seen outside-in. I noticed when GC collecting that there were other referrers to the PGResult - that may have to do with the problem. Could be a red herring.

**Comment by dvarrazzo** [comment]
Hello,

I tried to look better into this issue. While it's true that even closing the cursor the memory doesn't seem to be freed, this seems true only for a singe cursor:

Even leaving the closed connections and cursors around, with a loop like:

```python
conns = []
curs = []

while True:
    conn = psycopg.Connection.connect()
    try:
        with conn.cursor() as cur:
            cur.executemany(statement, items, returning=True)
            conn.commit()
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        conn.close()

    conns.append(conn)
    curs.append(cur)
```

The memory is reported to be stable during the iteractions. Showing it with [`mem_usage_ui`](https://github.com/parikls/mem_usage_ui):

![Image](https://github.com/user-attachments/assets/e6845843-d501-4133-a851-51a85c25d261)

Conversely, if the cursor is not closed, replacing the block above with:

```python
        cur = conn.cursor()
        cur.executemany(
            statement, items, returning=True
        )
        conn.commit()
```

then the memory grows indefinitely.

![Image](https://github.com/user-attachments/assets/237d8e3c-4cfd-491e-bdc5-47f44adc3108)

Playing with first a small set than a larger one, with the opposite, shows the memory going up but not down, and not growing unbounded. An example of the former:

![Image](https://github.com/user-attachments/assets/9dd4469e-4baa-49e9-ad9f-07f6277b41ec)

So, I don't see a leak in psycopg here, it seems that we are disposing of the unused objects correctly. I tend to think that the memory we see here is the one allocated by Python and never released.

I would like to close this issue here. If there are more information please let us know.
